src/serac/physics/thermal_conduction_functional.hpp

```diff
@@ -0,0 +1,481 @@
+        x_tensor[1] = x[1];
+      }
+      if constexpr (dim > 2) {
+        x_tensor[2] = x[2];
+      }
+
+      return temp(x_tensor, t);
+    });
+
+    bcs_.addEssential(temp_bdr, temp_bdr_coef_, temperature_);
+  }
+
+  /**
+   * @brief Advance the timestep
+   *
+   * @param[inout] dt The timestep to advance. For adaptive time integration methods, the actual timestep is returned.
+   */
+  void advanceTimestep(double& dt) override
+  {
+    temperature_.initializeTrueVec();
+
+    if (is_quasistatic_) {
+      nonlin_solver_.Mult(zero_, temperature_.trueVec());
+    } else {
+      SLIC_ASSERT_MSG(gf_initialized_[0], "Thermal state not initialized!");
+
+      // Step the time integrator
+      ode_.Step(temperature_.trueVec(), time_, dt);
+    }
+
+    temperature_.distributeSharedDofs();
+    cycle_ += 1;
+  }
+
+  template <typename MaterialType>
+  void setMaterial(MaterialType material)
+  {
+    K_functional_.AddDomainIntegral(
+        Dimension<dim>{},
+        [material]([[maybe_unused]] auto x, auto temperature) {
+          // Get the value and the gradient from the input tuple
+          auto [u, du_dx] = temperature;
+
+          auto flux = material.thermalFlux(u, du_dx);
+
+          auto source = u * 0.0;
+
+          // Return the source and the flux as a tuple
+          return serac::tuple{source, flux};
+        },
+        mesh_);
+
+    M_functional_.AddDomainIntegral(
+        Dimension<dim>{},
+        [material]([[maybe_unused]] auto x, [[maybe_unused]] auto temperature) {
+          auto [u, du_dx] = temperature;
+
+          auto source = material.cp * material.rho;
+
+          auto flux = 0.0 * du_dx;
+
+          // Return the source and the flux as a tuple
+          return serac::tuple{source, flux};
+        },
+        mesh_);
+  }
+
+  template <typename TemperatureType>
+  void setTemperature(TemperatureType temp_function)
+  {
+    // Project the coefficient onto the grid function
+    mfem::FunctionCoefficient temp_coef([temp_function](const mfem::Vector& x, double t) -> double {
+      tensor<double, dim> x_tensor;
+      x_tensor[0] = x[0];
+      if constexpr (dim > 1) {
+        x_tensor[1] = x[1];
+      }
+      if constexpr (dim > 2) {
+        x_tensor[2] = x[2];
+      }
+
+      return temp_function(x_tensor, t);
+    });
+
+    temp_coef.SetTime(time_);
+    temperature_.project(temp_coef);
+    gf_initialized_[0] = true;
+  }
+
+  template <typename SourceType>
+  void setSource(SourceType source_function)
+  {
+    K_functional_.AddDomainIntegral(
+        Dimension<dim>{},
+        [source_function, this]([[maybe_unused]] auto x, auto temperature) {
+          // Get the value and the gradient from the input tuple
+          auto [u, du_dx] = temperature;
+
+          auto flux = du_dx * 0.0;
+
+          auto source = source_function(x, time_);
+
+          // Return the source and the flux as a tuple
+          return serac::tuple{source, flux};
+        },
+        mesh_);
```

Comment:
`x` looks like it's always used, so its `[[maybe_unused]]` attribute can be removed, I think.

Here, the source function is a function of space and time, is it worthwhile to let it depend on temperature as well (I imagine space + time account for 90+ % of the uses).